first two articles

kevinfalting opened the pull request:
This adds some articles about failing safe and the negated happy path form.

View the rendered versions here: https://github.com/kevinfalting/kevinfalting.github.io/blob/kf/fail-safe/README.md

kevinfalting committed:
add negated-happy-path-form draft

kevinfalting committed:
add fail-safely draft

kevinfalting committed:
update readme with first two articles

negated-happy-path-form
fail-safely

also, some additional words and a link to linkedin.

sethetter commented on fail-safely.md:
```diff
@@ -0,0 +1,100 @@
+  case Active, Pending:
+    // okay, continue...
+
+  case Deleted:
```
If you couple this approach with unit tests that check the function with all possible values of the pseudo-enum, you can catch these before production runtime.

You’d have to maintain a list of all valid values, but that’s very doable. Or use iota and loop until you hit the max value.

kevinfalting replied:
Yeah, I removed a section I had about this because I felt like it distracted from the main idea. The difficult part is designing a test that will catch and fail for new values that aren't explicitly handled. You end up maintaining a duplicate set of these values for the test and if the function is one of those gnarly ones that require a ton of setup... 🤮 In some cases it starts to feel like testing a data structure. Of course, if an exhaustive test can be written, it's a good idea to do so.

The real solution is for there to be a language-level mechanic that can enforce this at compilation. Short of that, some linters claim to look for exhaustiveness; I haven't used them myself.

As the last line of defense, it's battle-tested in production, and fortunately, this pattern will throw errors that can point to exactly where the problem is.

sethetter replied:
> The real solution is for there to be a language-level mechanic that can enforce this at compilation.

Agreed! Having exhaustiveness checking is something I take for granted I suppose, working in TypeScript primarily for the last couple years 🙃

sethetter commented on fail-safely.md:
````diff
@@ -0,0 +1,100 @@
+}
+```
+
+And that's really it. Avoid implictly failing open, and beyond that choose the right tool for the job. When I'm dealing with a set of related values, I typically choose a switch block. It results in more lines of code, but I believe the trade off for explicitness and debuggability are worth it.
````
I’m really starting to wonder when implicitness is ever a good thing. I guess only when a piece of information is so commonly present in some specific context would explicitness add more noise than value. But otherwise, and especially in type definitions, the more explicit you can be, the safer your code.

kevinfalting replied:
I'd be willing to argue that the only form to always avoid is the implicit fail open.

> I guess only when a piece of information is so commonly present in some specific context would explicitness add more noise than value.

If by more noise you mean more text on screen, I disagree that it's noise. At a certain subjective point, a conditional is simpler to understand and change when it goes from an `if` block to a `switch` block. It's fine to start with the implicit fail closed form and upgrade to explicit.

One benefit of being explicit is that it's more likely that someone thought about it rather than a certain case being missed and it just happens to work.

But yes, it's common for these ideas to get feedback that they feel like more code than necessary. As long as folks are deliberately choosing one form over the other, I think that's a win.

sethetter replied:
> If by more noise you mean more text on screen, I disagree that it's noise.

I wasn't implying this was a scenario that created noise, I was considering implicit/explicit information about a program _in general_.

Memory allocation for example, since go is garbage collected, is fairly implicit in go programs, and that seems like a good thing. Though that has nothing to do with your business logic, so maybe that's the line.

Just thinking! We're aligned here. The `switch` encourages more a more explicit expression of the logic, which is a positive.

sethetter commented on fail-safely.md:
```diff
@@ -0,0 +1,100 @@
+
+And that's really it. Avoid implictly failing open, and beyond that choose the right tool for the job. When I'm dealing with a set of related values, I typically choose a switch block. It results in more lines of code, but I believe the trade off for explicitness and debuggability are worth it.
+
+I came to this after working in a large system that for years made the safe-at-the-time assumption that there would only ever be two user types. Then, along came a third and fourth. Unfortunately, introducing these new user types required the exercise of combing through all the execution paths and closing peices off - who knows what paths we've missed that are still implicitly failing open! Ideal ...
```
The last sentence there drives it home! By having to focus on the requirements of the new user type, you avoid having to consider the requirements of every _other_ user type and work in the negative. That reduces the chance of accidental behavior.

sethetter commented on fail-safely.md:
````diff
@@ -0,0 +1,100 @@
+
+```go
+switch user.State {
+  case Active, Pending:
````
Another technique I can imagine is having defined groups or categories of types that are defined once and referenced in relevant code paths.

Though even with that, the logic should fail closed outside of that defined group.

kevinfalting replied:
I did consider and play with this idea, but it feels like a DRYing effort, and conditional checks are better thought of as purpose-built for their context. While adding a layer of abstraction is possible, in most cases it's unnecessary.

The bit of exploration I did felt like I was abstracting a switch statement and/or a map. I'd be interested in seeing an example of what you're imagining if you have time. (no worries if not ❤️)

sethetter replied:
> While adding a layer of abstraction is possible, in most cases it's unnecessary.

Totally! In some cases it's helpful, depends on the context. I agree it's best to be cautious about reaching for abstraction too early.

sethetter commented on fail-safely.md:
````diff
@@ -0,0 +1,100 @@
+)
+```
+
+However, in code, I may find something like:
````
I think to better illustrate your idea here with the code sample, more of a specific example would help. Without knowing what the surrounding code is actually doing, it’s hard to know whether this is really a fail closed or fail open.

kevinfalting replied:
I explain what's wrong with it _after_ the code example, which is probably where the issue is. I'll rework this to explain what to look for ahead of showing the code example.

sethetter commented on negated-happy-path-form.md:
````diff
@@ -0,0 +1,121 @@
+if !(isThis || isThat) {}
+```
+
+Stop here, I believe I've made my point. It's a simple change and we're just getting progressively more silly from this point on.
````
```suggestion
Stop here, I believe I've made my point. It's a simple change and we're just getting progressively sillier from this point on.
```

It’s a word, I swear. My kids say it all the time.

sethetter commented on negated-happy-path-form.md:
````diff
@@ -0,0 +1,121 @@
+Depending on the complexity of the expression, I may go between the second and third forms, sometimes combining them.
+
+```go
+isThis := a && b
````
I love this style too because the var names add meaning to the otherwise confusing expression. Boolean conditions can be some of the hardest things for our brains to grok for some reason. Being able to break it down like this helps immensely.

sethetter commented:
Both great reads! Left some thoughts. Do with them what you will 😄 

I appreciate the conciseness. You really are a go dev 💯

kevinfalting committed:
add .gitignore

A .DS_store file showed up, this ignores it.

kevinfalting committed:
update the link name

kevinfalting committed:
add link to source repo for this site.

Folks can open issues and debate the content on github. Yeehaw.

kevinfalting committed:
add newline for source formatting

kevinfalting committed:
remove "sillier" section

It felt like the sillier second part of this article detracted from the
main point, so I've removed it. It was mostly about the topic of scoping
intermediate variables in conditional statements, and that can be a
separate article.

Additionally, I added an actual return statemtent to the last example to
keep it consistent with all the other examples.

kevinfalting committed:
adjustments for clarity and the narrative flow

kevinfalting committed:
don't shame iota, just don't bring it up

iota is not the right tool here. It has a specific set of things it's
good for, but it causes more headache than it helps in the scenarios
described in this article. I don't want to see it used, but there's no
reason I should shame it's usage, therefor I just won't mention it.

kevinfalting committed:
address feedback about fail-safely wording

1. don't mention enums, otherwise it sounds like that's the problem I'm
   trying to solve.
2. don't mention "ticket", non-enterprise folks _may_ not know what that
   means.
3. reword the last paragraph.

kevinfalting committed:
reword negated-happy-path-form article based on feedback

1. intro feels like i'm angry.
2. "clever" is a loaded term with negative connotations in this
   industry.
3. consider a cheeky double negative like concluding sentence.

kevinfalting committed:
remove stray 'is' from sentence
